`Hds::Form::TextInput::Field` and `Hds::Form::MaskedInput` - Refactor to use `Hds::Form::VisibilityToggle`

alex-ju opened the pull request:
:warning: This PR awaits to be released in the next major version :warning:

### :pushpin: Summary

- Add `Hds::Form::VisibilityToggle` component 
- Refactor `Hds::Form::MaskedInput` to use `Hds::Form::VisibilityToggle` (includes renaming `@isMasked` to `@isContentMasked`)
- Add `Hds::Form::VisibilityToggle` (controlled via `@hasVisibilityToggle`) to `Hds::Form::TextInput::Field` when `@type="password"`
    - `@hasVisibilityToggle` is set to `true` by default, which means the toggle is rendered by default

### :hammer_and_wrench: Detailed description

We took the following steps in order to achieve this in a systematic way:

 - we abstract the visibility toggle button from `MaskedInput` into a ~~utility~~ form base element component named `VisibilityToggle`
 - we use the `VisibilityToggle` in `TextInput::Field` (introducing it at the `TextInput::Based`-level would create side effects because of the need to have a wrapper element over the input control; it would also not benefit any of our consumers)

Note: the mechanism of toggling content visibility in a generic input (such as the ones generated by `MaskedInput`, namely `<input type="text">` and `<textarea>`) is different from the one used in `Hds::Form::TextInput::Field` for a `<input type="password">`. In the former implementation, we use a CSS modifier to style the content differently (obfuscating it) while in the latter we swap the `type` from `password` to `text` to make the most of the native password input implementation and semantics.

### 🧑‍💻 Impact on consumers

There are ~12 instances of `Hds::Form::TextInput::Field` with `@type="password"`, all of them in Cloud UI.
There are no instances of `Hds::Form::TextInput::Base` with `@type="password"`.

### :camera_flash: Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr><td>

<img width="339" alt="Screenshot 2023-08-23 at 10 16 29" src="https://github.com/hashicorp/design-system/assets/788096/dea0ef43-f528-40ef-a9d9-81c47854b74f">

</td><td>

<img width="334" alt="Screenshot 2023-08-23 at 10 16 48" src="https://github.com/hashicorp/design-system/assets/788096/8163bea0-edc8-4729-a9cf-1dfe133851bf">

</td></tr>
</table>


### :link: External links

<!-- Issues, RFC, etc. -->
Jira ticket: [HDS-2425](https://hashicorp.atlassian.net/browse/HDS-2425)

***

### 👀 Reviewer's checklist:

- [x] +1 Percy if applicable
- [x] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed
- [x] Confirm that A11y tests have been run locally for this component

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.


[HDS-2425]: https://hashicorp.atlassian.net/browse/HDS-2425?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

andgen404 commented on packages/components/addon/components/hds/form/masked-input/base.hbs:
```diff
@@ -21,14 +21,11 @@
-    class="hds-form-masked-input__toggle-button"
-    type="button"
-    aria-label={{this.ariaLabel}}
+  <Hds::VisibilityToggle
```
In figma we have this written as `_toggleVisability` just wanted to make sure there's alignment in naming.

KristinLBradley replied:
To me, "toggleVisability" sounds more like an action vs. an element which is used for performing the action.

heatherlarsen replied:
@andgen404 It looks like what @didoo is proposing is what we discussed today in our 1:1. That approach makes sense, so can you please move it to the Form Primitives page in Figma?

jorytindall reviewed (commented):
Left one comment that we need to get alignment on between code and Figma before releasing this.

jorytindall commented on packages/components/addon/components/hds/form/text-input/field.hbs:
```diff
@@ -14,16 +14,38 @@
-      id={{F.id}}
-      aria-describedby={{F.ariaDescribedBy}}
-    />
+    {{#if (eq @type "password")}}
```
Also, referencing [this thread](https://www.figma.com/file/sJHCJxIXKmRf9mlUbV1u0K?node-id=11530:27389&mode=design#534284915)

alex-ju replied:
Very good point here, Jory. I think there were assumptions on both parts.

> I don't have a strong preference on which route we go; more flexibility can be good, but simplifying the experience for consumers would also bring consistency.
Same here, no strong opinions, I can see the tradeoffs. I admit that I've not considered introducing this as an optional item. We'll then have to think about what would be the name of the boolean argument and what would be the default 🤔

didoo replied:
+1 on considering how to allow consumers to decide if have the toggle (default) or not (via an option)

heatherlarsen replied:
I would lean into our principle of "guidance over control" and provide flexibility by making it optional. Would it be a breaking change if the default value was set to `true`?

alex-ju replied:
Added a `hasVisibilityToggle` argument (defaulting to `true`) that controls this feature – @andgen404 would this loosely match what you have in Figma?

andgen404 replied:
@alex-ju Yes, that would match what I have in Figma

didoo commented on packages/components/addon/components/hds/form/text-input/field.js:
```diff
@@ -0,0 +1,19 @@
+import { action } from '@ember/object';
+
+export default class HdsFormTextInputFieldComponent extends Component {
+  @tracked isMasked = this.args.isMasked ?? true;
```
Alternatives https://www.thesaurus.com/browse/readable:
- `plain`
- `clear`

didoo commented on packages/components/addon/components/hds/visibility-toggle/index.js:
```diff
@@ -0,0 +1,30 @@
+   * @type {string}
+   * @default 'Show masked content'
+   */
+  get ariaLabel() {
```
See comment above about being more agnostic, as component

alex-ju commented:
@didoo I hope I addressed all the above feedback. I've not yet updated the docs website – I will add a commit as soon as we settle on arguments naming

didoo commented on packages/components/addon/components/hds/form/masked-input/base.hbs:
```diff
@@ -21,16 +21,14 @@
+    @ariaMessageText={{this.ariaMessageText}}
     aria-controls={{this.id}}
+    class="hds-form-masked-input__toggle-button"
     {{on "click" this.onClickToggle}}
```
[nitpick] for consistency with `onClickTogglePasswordReadability`?
```suggestion
    {{on "click" this.onClickToggleMasking}}
```

didoo commented on packages/components/addon/components/hds/form/text-input/field.js:
```diff
@@ -0,0 +1,59 @@
+    if (this.args.visibilityToggleAriaMessageText) {
+      return this.args.visibilityToggleAriaMessageText;
+    } else if (this.isPasswordReadable) {
+      return 'Password is now visible';
```
PS: in case you make the change, don't forget to do the same in the `MaskedInput` too

didoo commented on packages/components/tests/dummy/app/templates/components/form/base-elements.hbs:
```diff
@@ -576,4 +576,23 @@
+
+  <Shw::Text::H2>Visibility toggle</Shw::Text::H2>
+
+  <Shw::Grid @columns={{4}} {{style width="fit-content"}} as |SG|>
```
you can probably use a flex container here, so no need to set the width and the columns

didoo reviewed (commented):
Added more comments (sorry, on a second pass other things stood out).

Re, the naming of components/variables/states (`VisibilityToggle`, `isVisible`, `isMasked`, `isPasswordReadable`, `showVisibilityToggle`) I don't see much we can do. The only thing that comes to mind is to rename `isMasked` in the `Masked::Input` to `isContentMasked` and then the current `isPasswordReadable` to `isPasswordMasked` so that we have for both:
```
@isVisible={{this.isContentMasked}}
@isVisible={{(this.isPasswordMasked)}} // instead of using a "not"
```
but not a blocker, just an idea

alex-ju commented on packages/components/app/styles/components/form/visibility-toggle.scss:
```diff
@@ -0,0 +1,26 @@
+$hds-form-visibility-toggle-button-size: 24px;
+
+.hds-form-visibility-toggle {
+  position: absolute;
```
Agree, the toggle itself should be agnostic of where it's placed within other components.

alex-ju commented on packages/components/addon/components/hds/form/masked-input/base.hbs:
```diff
@@ -21,16 +21,14 @@
-    aria-label={{this.ariaLabel}}
+  <Hds::Form::VisibilityToggle
+    @isVisible={{this.isMasked}}
+    @ariaLabel={{this.ariaLabel}}
```
Indeed, I added them in this PR

alex-ju commented on packages/components/addon/components/hds/form/text-input/field.js:
```diff
@@ -0,0 +1,59 @@
+import { action } from '@ember/object';
+
+export default class HdsFormTextInputFieldComponent extends Component {
+  @tracked isPasswordReadable = this.args.isPasswordReadable ?? false;
```
Good thinking, keeping it as an internal variable helps avoid this state

alex-ju commented:
> Added more comments (sorry, on a second pass other things stood out).
> 
> Re, the naming of components/variables/states (`VisibilityToggle`, `isVisible`, `isMasked`, `isPasswordReadable`, `showVisibilityToggle`) I don't see much we can do. The only thing that comes to mind is to rename `isMasked` in the `Masked::Input` to `isContentMasked` and then the current `isPasswordReadable` to `isPasswordMasked` so that we have for both:
> 
> ```
> @isVisible={{this.isContentMasked}}
> @isVisible={{(this.isPasswordMasked)}} // instead of using a "not"
> ```
> 
> but not a blocker, just an idea

I think your suggestion makes sense. It would technically be a breaking change so I bumped this to a major, but we have at least two other PRs waiting for a major release, so I'm not concerned about making this change here and now.

I hopefully addressed the feedback from the second round and updated the docs to expose the arguments that are meant to be public.

didoo commented on packages/components/addon/components/hds/form/masked-input/field.hbs:
```diff
@@ -17,7 +17,7 @@
     <Hds::Form::MaskedInput::Base
```
I _think_ we should also pass down the `@visibilityToggleAriaLabel` and `@visibilityToggleAriaMessageText`, right?

alex-ju replied:
indeed, pushed a commit to forward them and added tests to cover these

didoo commented on website/docs/components/form/masked-input/partials/code/component-api.md:
```diff
@@ -47,7 +53,7 @@ The Masked Input component has two different variants with their own APIs:
     Input control’s value.
   </C.Property>
-  <C.Property @name="isMasked" @type="boolean" @default="true">
+  <C.Property @name="isContentMasked" @type="boolean" @default="true">
```
I _think_ we should document also the `@visibilityToggleAriaLabel` and `@visibilityToggleAriaMessageText` (passed down to the `::Base` element)

didoo commented on packages/components/addon/components/hds/form/text-input/base.js:
```diff
@@ -57,6 +57,11 @@ export default class HdsFormTextInputBaseComponent extends Component {
       classes.push(`hds-form-text-input--is-invalid`);
     }
 
+    // add a class based on the @hasVisibilityToggle argument
```
Alternatively, we could call this `@hasExtraPaddingRight` (or something like this), which makes it more agnostic, and leave it here.

didoo replied:
OK, never mind: I see we also have a `hds-form-text-input__visibility-toggle` class, so let's keep it as is. We can generalize in the future if necessary.

didoo commented on packages/components/tests/acceptance/percy-test.js:
```diff
@@ -161,6 +161,9 @@ module('Acceptance | Percy test', function (hooks) {
     await visit('/utilities/dismiss-button');
     await percySnapshot('DismissButton');
 
+    await visit('/utilities/visibility-toggle');
+    await percySnapshot('VisibilityToggle');
+
```
Not needed anymore.
```suggestion
```

didoo commented on packages/components/tests/dummy/app/router.js:
```diff
@@ -76,6 +76,7 @@ Router.map(function () {
     this.route('dismiss-button');
     this.route('interactive');
     this.route('menu-primitive');
+    this.route('visibility-toggle');
```
Not needed anymore.
```suggestion
```

didoo commented on packages/components/tests/dummy/app/templates/components/form/text-input.hbs:
```diff
@@ -215,6 +232,47 @@
     </SG.Item>
   </Shw::Grid>
 
+  <Shw::Text::H3>Password</Shw::Text::H3>
```
Now that I've added the `Types` above, do you think this section becomes redundant?

alex-ju replied:
Indeed, they feel redundant now so I removed the 'Password' section in the showcase

didoo commented on website/docs/components/form/text-input/partials/code/component-api.md:
What do you think of adding an example, in the "How to use" section, about disabling the visibility toggle for a password field? It's an indirect way to tell consumers that this "thing" exists now, and that by default is rendered (but can be turned off)

alex-ju replied:
Added a section for "Password" input fields, showcasing the default and how to use the `@hasVisibilityToggle` argument

didoo reviewed (commented):
Added a few more comments

didoo reviewed (approved):
Good to go! 🚢

(it was quite a journey :) )
